fix extrainfo error & holdout columns during update_motifs

snystrom opened the pull request:
Closes #17.

snystrom committed:
Hold out poorly supported extrainfo types & remerge.

Addresses #13. https://github.com/bjmt/universalmotif/issues/13#issuecomment-803366926

snystrom committed:
extra checks for safety

snystrom committed:
update holdout behavior

snystrom committed:
add todo & more safety checks

snystrom committed:
fixes extrainfo error with 1 column. Closes #17.

snystrom committed:
Use name & consensus to merge heldout cols

snystrom committed:
Add unit tests for holdout behavior.

snystrom committed:
Merge branch 'update-motif-holdout'

snystrom committed:
version bump for holdout feature

snystrom commented:
Realized I screwed up with the github PR so moving the text from #18 here since that & #17 are addressed in this PR by accident.

This change addresses issues discussed in #13 (https://github.com/bjmt/universalmotif/issues/13#issuecomment-803366926) related to how complex metadata are stored in extrainfo. This solution ignores all data types that are not character, integer, or numeric when calling update_motifs in order to prevent munging the data stored in those columns. They are then joined back to the updated motif_df currently by merging on name and consensus. Moving forward this should probably be made into a more unique identifier. Maybe just the row number of the input data.frame, but I wasn't 100% certain sort order would be preserved in all cases. If so, that's an easy swap (see the id_cols variable).

I also added a few unit tests to ensure these behave properly.

bjmt commented on R/universalmotif_df.R:
```diff
@@ -141,12 +141,27 @@ update_motifs <- function(motif_df, extrainfo = FALSE) {
+      # TOOD: hold out unsupported datatypes
+      # Current "supported" types are "character", "numeric" and "integer"
+      # Should logical be held out or not?
+      extrainfo_types <- vapply(extrainfo_new, class, character(1))
```
Can you use `typeof` instead of `class`? I think the former should be safer.

snystrom replied:
If I use `typeof` this interprets factors as numeric, not factors, so they get converted to their integer values (the new unit tests catch this though, so there's at least the infrastructure to catch when it's fixed). I guess another option is to handle factors separately? I'm not sure how to handle this one, so I've left it as `class` for now.

bjmt commented on R/universalmotif_df.R:
```diff
@@ -201,7 +216,14 @@ update_motifs <- function(motif_df, extrainfo = FALSE) {
+    # Add back any heldout info
+    new_df <- to_df(m, extrainfo)
+    # TODO: update id_cols strategy
+    return(merge(new_df, extrainfo_holdouts, by = id_cols, all.x = TRUE))
```
Can you wrap the returning object with `structure(..., class = c("universalmotif_df", "data.frame"))` here?

snystrom replied:
Whoops! Forgot to do that. Fixed. 

I also added a new unit test to catch similar errors in the future.

bjmt commented:
Found a weird bug. Haven't investigated into why it's happening, I'll try to look into it when I have a bit more time later. The following code generates duplicate rows:

```r
m <- create_motif()
m <- c(m, m)
m[[1]]["extrainfo"] <- c(A = "2")
m[[2]]["extrainfo"] <- c(B = "4")
update_motifs(to_df(m, TRUE), TRUE)
```

snystrom commented:
> 
> 
> Found a weird bug. Haven't investigated into why it's happening, I'll try to look into it when I have a bit more time later. The following code generates duplicate rows:
> 
> ```r
> m <- create_motif()
> m <- c(m, m)
> m[[1]]["extrainfo"] <- c(A = "2")
> m[[2]]["extrainfo"] <- c(B = "4")
> update_motifs(to_df(m, TRUE), TRUE)
> ```

Ah, yeah, I know why this happens. When converting from motif -> df, each named `extrainfo` object becomes a column, so:

```
held_out <- data.frame(name = c("motif1", "motif1"),
                 A = c("2", NA)
                 B = c(NA, "4"))
```

During the update this gets joined using `base::merge` which will make duplicate rows if the extrainfo has multiple row matches (which it will because of the identical names). I think the fix is to use a unique row identifier instead of motif name, which is not guaranteed to be unique.

Is row order guaranteed to remain stable during this transformation? If it is, we can use the row number as the id.

snystrom committed:
add failing unit test for row duplication bug

snystrom committed:
Drop the id_col approach & use cbind instead.

Fixes duplicate row issues caused by merge.

snystrom committed:
add check for universalmotif_df after update

snystrom committed:
Added notes RE: typeof vs class in holdouts detection

snystrom commented:
I fixed the row duplication bug by reverting the `id_cols` strategy and using `cbind` instead. This assumes row order is always preserved. I think that is always the case, but not sure if there's something deep in the internals which could cause that issue. The unit tests should catch it though and they pass.

I also fixed the issues you raised above. The only outstanding thing is the `typeof` vs `class` issue.

bjmt commented:
Thanks! It's looking good. As far as I can remember the order of motifs is never changed anywhere, internally or externally. Using `cbind` should be safe.

With regards to the `typeof` vs `class`: you're right, I had forgotten about factors. However, adding `x <- vapply(x, as.character, character(1))` inside the `clean_up_extrainfo_df()` function (bottom of the file) seems to properly coerce factors to characters.

Also, I noticed your comment about logical columns: I think it's fine to include those in the types that can be coerced into characters for the `extrainfo` slot. However this had me thinking about these character coercions occurring silently, which I'm not sure I like. How about a message stating "Note: non-character extrainfo columns coerced to character [column1, column2, column3]" or something? Also I think it would be a good idea to also have a message for when these holdout columns are detected as well.

(While your at it, feel free to change the "discarding unknown slots" warning to a message as you mentioned in #13. Otherwise I can change it after merging.)

snystrom committed:
Swap warning -> message for discarding slots

snystrom committed:
Added force parameter to update_motifs and to_list

snystrom commented:
> With regards to the `typeof` vs `class`: you're right, I had forgotten about factors. However, adding `x <- vapply(x, as.character, character(1))` inside the `clean_up_extrainfo_df()` function (bottom of the file) seems to properly coerce factors to characters.

I tried this out and am not quite happy with the behavior. I'll discuss more below, but the relevant solution here is that I added additional filtering to use `typeof` for all types that are not factors, which filter using `class`. In this way, factors and their levels can be preserved if just updating a motif_df. I wrote it so this strategy can be expanded to similar `class`/`type` mismatches in the future if something important comes up.

> Also, I noticed your comment about logical columns: I think it's fine to include those in the types that can be coerced into characters for the `extrainfo` slot. However this had me thinking about these character coercions occurring silently, which I'm not sure I like. How about a message stating "Note: non-character extrainfo columns coerced to character [column1, column2, column3]" or something? Also I think it would be a good idea to also have a message for when these holdout columns are detected as well.

I agree that I don't like the silent coercion, especially since you might want to `update_motifs` just to change the `motif` name, while keeping on with the manipulation of other data columns where type may be important. For this reason, I think it's important to prevent data munging unless the user explicitly wants it. To address this, I've added a third parameter `force` (default: `FALSE`) to `update_motifs()` and `to_list()` which when set to `TRUE` does not hold out any columns and forces them to characters into the `extrainfo` (in other words, `update_motifs(df, TRUE, TRUE)` gives the original behavior before the changes in this PR). This also has the side effect of destructively changing the col type in the data.frame when calling `update_motifs(df, TRUE, TRUE)`. Because the user has to explicitly ask for this behavior, I opted not to write a `message`, but it could be done. I tried to be explicit in the docstring that this operation is destructive. I added some unit tests for this as well.

Because of this change, what do you think about changing the `extrainfo` default to `TRUE`? I don't have a strong opinion one way or the other, but I don't see any downsides to it now that data are not munged and odd data-structures aren't carried over by default. It might make things feel more pleasant to use. 

> (While your at it, feel free to change the "discarding unknown slots" warning to a message as you mentioned in #13. Otherwise I can change it after merging.)

Done! I left the other warnings as is for now, although `Restoring missing slot(s)` could probably also be a message. I updated the unit tests to `expect_message` or `suppressMessages` where needed to reflect this change.

bjmt commented:
> I tried this out and am not quite happy with the behavior. I'll discuss more below, but the relevant solution here is that I added additional filtering to use typeof for all types that are not factors, which filter using class. In this way, factors and their levels can be preserved if just updating a motif_df. I wrote it so this strategy can be expanded to similar class/type mismatches in the future if something important comes up.

Can you expand on this a bit more? I think I'm misunderstanding the problem you're seeing, because when I make the change to `clean_up_extrainfo_df()` I mentioned earlier alongside using solely `typeof` everything works fine for me:
```r
...
extrainfo_types <- vapply(extrainfo_new, typeof, character(1))
extrainfo_holdout_cols <- extrainfo_types %in% c("character", "numeric", "integer")
extrainfo_holdouts <- extrainfo_new[, extrainfo_holdout_cols, drop = FALSE]
...
```
```r
clean_up_extrainfo_df <- function(x) {
  y <- names(x)
  x <- vapply(x, as.character, character(1))
  x <- as.character(x)
  if (all(is.na(x))) character()
  else {
    names(x) <- y
    x[!is.na(x)]
  }
}
```
Perhaps I'm missing where you refer to having discussed this more.

> To address this, I've added a third parameter force (default: FALSE) to update_motifs() and to_list() which when set to TRUE does not hold out any columns and forces them to characters into the extrainfo (in other words, update_motifs(df, TRUE, TRUE) gives the original behavior before the changes in this PR). This also has the side effect of destructively changing the col type in the data.frame when calling update_motifs(df, TRUE, TRUE). Because the user has to explicitly ask for this behavior, I opted not to write a message, but it could be done. I tried to be explicit in the docstring that this operation is destructive. I added some unit tests for this as well.

Good idea!

> Because of this change, what do you think about changing the extrainfo default to TRUE?

Thinking back on my playing about so far, I definitely set this to `TRUE` more often than not. Seems like a better choice.

One final thing: the "Discarding extra slots" message is now showing when I run `update_motifs(mydf, extrainfo = TRUE)` on a `universalmotif_df` with extra columns (but not for `to_list(mydf, extrainfo = TRUE)`). 

Thanks for being patient with my slow replies and endless nagging.. I think we're getting close now.

snystrom committed:
Update unit tests for explicit extrainfo & check messages

snystrom committed:
set extrainfo = TRUE as new default.

snystrom commented:
> Can you expand on this a bit more? I think I'm misunderstanding the problem you're seeing, because when I make the change to `clean_up_extrainfo_df()` I mentioned earlier alongside using solely `typeof` everything works fine for me:
>   `x <- vapply(x, as.character, character(1))`
> 
> Perhaps I'm missing where you refer to having discussed this more.

Yeah, I didn't fully explain myself there. Using this approach causes lists, logicals, and factors to get destroyed during an update (for an explicit example, the unit tests checking equality for logicals, lists & factors fail using this strategy: lists & logical columns aren't propagated to the updated df, while factors become chars). Factors and logicals really aren't useful inside a motif (which is why they should become characters during forced coercion), but they're really useful inside a data.frame, so I think it's best to not munge them if it can be avoided. Same thing for list columns, which I frequently abuse, so I'm partial to them working.

> Thinking back on my playing about so far, I definitely set this to `TRUE` more often than not. Seems like a better choice.

Me too. I've set `extrainfo = TRUE` as the default for `update_motifs()` , `to_list()`, and `requires_update()`. I updated the unit tests to set `extrainfo` explicitly if it wasn't already to preserve expected behavior.

> One final thing: the "Discarding extra slots" message is now showing when I run `update_motifs(mydf, extrainfo = TRUE)` on a `universalmotif_df` with extra columns (but not for `to_list(mydf, extrainfo = TRUE)`).

Huh. I don't replicate that on my machine. I added some units tests to catch this though, so if the issue persists on your end maybe that will help debug?

> Thanks for being patient with my slow replies and endless nagging.. I think we're getting close now.

Absolutely! This is how we ship good features! I really appreciate the careful feedback.

bjmt commented:
> Yeah, I didn't fully explain myself there. Using this approach causes lists, logicals, and factors to get destroyed during an update (for an explicit example, the unit tests checking equality for logicals, lists & factors fail using this strategy: lists & logical columns aren't propagated to the updated df, while factors become chars). Factors and logicals really aren't useful inside a motif (which is why they should become characters during forced coercion), but they're really useful inside a data.frame, so I think it's best to not munge them if it can be avoided. Same thing for list columns, which I frequently abuse, so I'm partial to them working.

I understand now. Let's just stick to your original `class` implementation then. No need to overcomplicate things if my `typeof` suggestion isn't a direct drop-in replacement (unless you can think of a reason to use both checks). 

However when I run `update_motifs(mydf, TRUE, TRUE)` I notice factor columns aren't probably being coerced to characters and show up as the underlying numbers. My `clean_up_extrainfo_df()` addition fixes that for me, but maybe you have a different fix in mind and/or are not seeing this problem?

> Huh. I don't replicate that on my machine. I added some units tests to catch this though, so if the issue persists on your end maybe that will help debug?

Not sure what changed but it's no longer happening now!

snystrom committed:
add unit tests to ensure force converts properly to char

snystrom committed:
Revert typeof holdout & use vapply

snystrom committed:
Change restoring slots warning to message.

snystrom commented:
> However when I run `update_motifs(mydf, TRUE, TRUE)` I notice factor columns aren't probably being coerced to characters and show up as the underlying numbers. My `clean_up_extrainfo_df()` addition fixes that for me, but maybe you have a different fix in mind and/or are not seeing this problem?

Oh no! Looks like I didn't check that carefully enough. I added some unit tests to catch this behavior in the future (for factors and some other data types). In light of this, I revisited the `vapply(x, as.character, character(1)` strategy in conjunction with the `class` strategy and that actually works fine. I think initially I was only trying 1 but not the other, or some weird combination which is why I never got it working. Regardless, using this approach all the unit tests pass, lists aren't munged, etc. So I removed the `typeof` bit & added the `vapply` bit to `clean_up_extrainfo_df()`. 

Finally, I went ahead and changed the "restoring missing slots" warning to a message & altered the unit tests accordingly. Feel free to revert that if you disagree. But since the columns are NA and aren't propagated unless filled, I figure it doesn't really constitute a warning. Plus, this way it can be safely suppressed when developing with these features without a lot of `tryCatch`ing.

bjmt commented:
Awesome! Everything looks good to me. Can't think of anything else. Thanks again! I really appreciate all the time you took to work on this.

snystrom commented:
Great! I think this also closes #13. Really excited about this new functionality. I think it's going to be a really powerful data structure.
